6주차 - 레스토랑 상세 페이지 구현하기

yxxnhx opened the pull request:
# 목표
레스토랑 상세 페이지 구현하기

### 구현사항
- [x] Home, About, Restaurants 관심사 분리하여 컴포넌트 만들기
- [x] 각각 페이지 라우팅하기
- [x] 레스토랑 상세 페이지 연결하기

### 진행사항
현재 Home, About, Restaurants 페이지를 각각 라우팅해놓은 상태입니다.
레스토랑의 상세 페이지를 보여줄 수 있도록 어떻게 할지 고민하고 있습니다.
차근차근히 해보도록 하겠습니다👊

yxxnhx committed:
라우팅을 위해 react-router-dom 설치

페이지 라우팅을 위하여 react-router-dom 설치하였습니다

yxxnhx committed:
웹 접속 시 index.html로 연결 설정

모든 웹 접속 시 index.html로 연결될 수 있도록 webpack에 devServer를 설정해주었습니다

yxxnhx committed:
index 컴포넌트 라우팅 설정

개별 페이지 세부 테스트를 위하여 index에 BrowserRouter 설정하였습니다

yxxnhx committed:
레스토랑 관련 관심사 RestaurantsPage로 분리

기존 App 컴포넌트에 있던 레스토랑 관련 관심사는 RestaurantsPage로 분리하였습니다

yxxnhx committed:
App 컴포넌트 라우팅 설정 및 각각 컴포넌트 연결

routes, route를 이용하여 라우팅 기본 틀을 잡았습니다

yxxnhx committed:
HomePage 컴포넌트 생성 및 about, restaurnats 페이지 연결

홈페이지 컴포넌트를 생성 후 link를 이용하여 about과 restaurants 페이지를 연결하였습니다

yxxnhx committed:
HomePage 테스트 파일 생성 및 기본 틀 작성

HomePage 테스트 파일을 생성하여 현재 홈페이지의 랜더링 여부를 테스트 하였습니다

yxxnhx committed:
AboutPage 컴포넌트 생성 및 틀 구현

about 페이지 컴포넌트 생성 후 기본 틀을 구현하였습니다

yxxnhx committed:
AboutPage 테스트 파일 생성 및 기본 틀 구현

어바웃 페이지가 잘 랜더링되는지 여부 확인하는 테스트 기본 틀 작성하였습니다

yxxnhx committed:
NotFoundPage 컴포넌트 생성 및 구현

잘못된 경로로 접속 시 보여질 페이지를 생성하였습니다

yxxnhx committed:
NotFoundPage 테스트 파일 생성 및 구현

NotFoundPage가 잘 랜더링되는지 여부를 테스트하였습니다

yxxnhx committed:
App 테스트 파일 생성 및 올바른 경로 시 보이는 페이지 여부 테스트

올바른 경로로 진입 시, 각각의 경로에 따른 페이지가 보일 수 있도록 테스트를 작성하였습니다

yxxnhx committed:
RestaurantsPage 테스트 조건 수정

테스토용 데이터에 있는 식당 이름으로 조건을 수정하여 테스트 통과하였습니다

yxxnhx committed:
HomePage 테스트 라우팅 설정

MemoryRouter로 라우팅 설정하여 테스트 내에서도 link 사용할 수 있도록 하였습니다

yxxnhx committed:
App 테스트 조건 수정

테스트용 데이터에 있는 데이터로 조건 수정하여 테스트 통과하였습니다

yxxnhx committed:
App 컴포넌트 import 경로  수정

AboutPage 컴포넌트 import 경로 에러로 인한 문제 경로 수정하여 해결하였습니다

yxxnhx committed:
App 테스트 /about 경로일 때  조건 수정

AboutPage가 랜더될 시에 나오는 문구 조건을 AboutPage 컴포넌트에 맞게 수정하였습니다

daadaadaah reviewed (changes requested):
항상 PR 일찍 남겨주시는 것 같은데, 좋은 프로세스인 것 같아요!👍
계속 진행해주세요🚀

yxxnhx committed:
App 컴포넌트에 RestaurantInfoPage 연결

route를 이용하여 상세페이지를 연결하였습니다

yxxnhx committed:
RestaurantContainer에 상세 페이지 주소 연결

각각의 레스토랑 상세 페이지 정보 연결을 위해 Link를 이용하여 연결하였습니다

yxxnhx committed:
RestaurantsContainer 테스트 통과를 위해 MemoryRouter 추가

컴포넌트에 Link 사용을 위해 MemoryRouter를 추가하여 에러를 해결하였습니다

yxxnhx committed:
RestauarantsPage 테스트 파일에 라우터  에러 문제 해결을 위해 MemoryRouter 추가

MemoryRouter를 이용하여 router 에러 해결을 하였습니다

yxxnhx committed:
RestaurantsInfoPage 컴포넌트 생성

loadRestaurantsInfo에 필요한 데이터의 아이디를 넣어서 상세 페이지가 출력될 수 있도록 하였습니다

yxxnhx committed:
레스토랑 상세 정보를 위해 actions에 loadRestaurantInfo를 추가

각각의 상세 정보를 받아올 수 있도록 fetchRestaurant에 restaurantId를 넣어서 디스패치 시킬 수 있도록 설정하였습니다

yxxnhx committed:
레스토랑 정보를 받아오기 위하여 reducer에 setRestaurantInfo를 추가

레스토랑 정보를 받아오는 액션을 위하여 reducer에 처리해주었습니다

yxxnhx committed:
테스트 에러 해결을 위하여 reducer 테스트 기본값에 restaurantInfo 설정

테스트 에러 해결을 위하여 기본값 null로 설정해주었습니다.

yxxnhx committed:
레스토랑 상세 정보 데이터를 받아오기 위하여 api에 fetchRestaurantInfo 추가

restaurantId를 넘겨받아 해당 데이터를 보여줄 수 있도록 주소 연결하였습니다

yxxnhx committed:
RestaurantInfoContaienr 컴포넌트 생성 및 구현

restaurantInfo를 넘겨받아 각각의 데이터를 구현할 수 있도록 하였습니다

yxxnhx commented:
# 진행사항
현재 RestaurantsInfoPage를 생성하여 데이터를 관리하고 각각의 데이터를 RestaurantsInfoContainer에서 보여줄 수 있도록 하였습니다

actions
```jsx
export function setRestaurantInfo(restaurantInfo) {
  return {
    type: 'setRestaurantInfo',
    payload: { restaurantInfo },
  };
}

/* 생략 */

export function loadRestaurantsInfo(restaurantId) {
  return async (dispatch) => {
    const restaurantInfo = await fetchRestaurantInfo(restaurantId);

    dispatch(setRestaurantInfo(restaurantInfo));
  };
}
```
reducer
```jsx
  setRestaurantInfo(state, { payload: { restaurantInfo } }) {
    return {
      ...state,
      restaurantInfo,
    };
  },
```

RestaurantsInfoPage
```jsx
  const { id } = useParams();

  useEffect(() => {
    dispatch(loadRestaurantsInfo(id));
  }, []);

  const restaurantInfo = useSelector((state) => ({
    restaurantInfo: state.restaurantInfo,
  }));
```

<img width="860" alt="image" src="https://user-images.githubusercontent.com/50559373/202427614-fb49a201-6530-412a-a44e-6165900efd34.png">

그런데 지금 restaurantInfo를 처음에는 null을 받다가 뒤늦게서야 데이터를 받아와 지금처럼 콘솔에 세 개가 찍힙니다ㅠㅠ
왜 그런 것일까요?

daadaadaah reviewed (changes requested):
1. 테스트 코드를 먼저 작성해보는 걸 의도적으로 훈련해보시는 걸 추천드립니다! 지금 질문 주신 내용이 테스트 코드를 먼저 작성했으면 해결될 수도 있는 이슈 같거든용!😊

## 질문에 대한 답변
RestaurantsInfoPage에서 `restaurantInfo`를 디스트러처링해서 뽑아오시고, 
```js
const { restaurantInfo } = useSelector((state) => ({
    restaurantInfo: state.restaurantInfo,
  }));
```
RestaurantsInfoContainer에서 restaurantInfo 가 없을 때에 대한 로직을 추가해주시면 될거예요!
```
  if (!restaurantInfo) {
    // ~~~
  }
```

yxxnhx committed:
RestaurantsInfoPage 컴포넌트 관심사 분리

페이지에서는 loadRestaurantsInfo에 아이디만 넘기고, 필요한 데이터를 가져와서 랜더링 여부는 컨테이너에서 할 수 있도록 관심사를 분리하였습니다

yxxnhx committed:
RestaurantsInfoContainer 컴포넌트  관심사 분리

필요한 데이터를 컨테이너에서 직접 관리할 수 있도록 관심사 분리 및 데이터가 없을 시에 로딩이 나올 수 있도록 조건을 추가하였습니다

yxxnhx committed:
테스트용 데이터 restaurantInfo 생성

fixtures에 테스트용 데이터 restaurantInfo를 추가하였습니다

yxxnhx committed:
RestaurantsInfoPage 테스트 파일 생성 및 구현

레스토랑 데이터 여부에 따른 테스트 조건을 생성하였습니다

yxxnhx committed:
RestaurantsInfoContainer 테스트 파일 생성 및 구현

데이터 여부에 따른 조건을 추가하여 테스트를 구현하였습니다

yxxnhx committed:
reducer 테스트 setRestaurantInfo 추가

레스토랑 정보를 가져오는 조건을 추가하여 setRestaurantInfo 테스트를 하였습니다

yxxnhx committed:
actions 테스트에 loadRestaurantsInfo 추가

loadRestauransInfo를 실행하면 setRestaurantInfo를 실행 여부를 테스트 하였습니다

yxxnhx committed:
테스트용 api fetchRestaurantInfo 추가

api를 테스트할 수 있도록 테스트용 fetchRestaurantInfo를 추가하였습니다

yxxnhx committed:
api 테스트에 fetchRestaurantInfo 추가

fetchRestaurantInfo를 실행시키면 restaurantInfo를 호출하여 동일한지 테스트하였습니다

yxxnhx committed:
eslint로 api 코드 수정

E2E 통과를 위해 api 코드 수정하였습니다

yxxnhx committed:
App 컴포넌트에 잘못된 접근 시 NotFoundPage 연결

잘못된 경로로 접근시 NotFoundPage로 연결될 수 있도록 수정하였습니다

yxxnhx committed:
App 테스트에 올바르지 않은 경로를 가지고 있을 시 조건 추가

올바르지 않은 경로로 접근 시에는 NotFoundPage 컴포넌트로 연결되는 것으로 테스트 조건을 추가하였습니다

yxxnhx commented:
# 진행사항
구조분해할당을 하여 restaurantInfo를 가져오고 restaurantInfo가 없을 경우의 조건을 추가하여 목록 조회까지 완료시켰습니다!
테스트 코드를 먼저 작성하면 방지할 수 있다니.. 틀을 간단히 잡고 테스트 코드를 짜야지 하다가 보니 이렇게 되었었네요ㅠㅠ
꼭 다음 과제 때는 의도적으로라도 테스트 코드를 먼저 작성하면서 잡을 수 있는 오류들은 잡아나가보겠습니다!

### 질문사항
![image](https://user-images.githubusercontent.com/50559373/202736772-3dcb2666-aec1-45c0-b1fd-79752b05bb75.png)

지금 현재 테스트 커버리지는 모두 통과를 하였고 레스토랑과 레스토랑 정보도 정상적으로 잘 출력이 되는데 E2E 테스트가 위와 같이 restaurantDetail이 나오지 않는다고 에러가 뜹니다ㅠㅠ
왜 그런 것일까요?
어떤 부분을 수정해야 이 부분이 해결이 될까요?

yxxnhx committed:
RestaurantInfoPage 컴포넌트에서 useEffect가 계속해서 랜더링되는 이슈 해결

디팬던시를 추가하여 계속해서 재랜더링되지 못하도록 해결하였습니다

yxxnhx committed:
restaurants/1로 접근 시 불가능한 이슈 해결

main.js를 import해오는 과정에서 404 not Found 에러가 발생하여 계층구조에 따라 /를 추가하여 문제 해결하였습니다

daadaadaah commented on src/reducer.test.js:
```diff
@@ -63,22 +63,31 @@ describe('reducer', () => {
 
+  describe('setRestaurantInfo', () => {
+    it('changes restaurantInfo', () => {
+      const initialState = {
+        restaurantInfo: [],
+      };
```
`restaurantInfo` 객체로 관리하는데, 이부분도 수정해줘야 하지 않을까여?ㅎㅎ

yxxnhx replied:
앗 그렇군요! null로 수정하겠습니다~

daadaadaah reviewed (changes requested):
1. presentational component 를 만들지 않고 구현하신 이유가 궁금합니다!
2. 질문 주신거 해결되셨나요?

yxxnhx committed:
reducer 테스트의 restaurantInfo 초기값 통일

기존에 null로 만들어둔 초기값을 빈배열로 잘못 설정하여 해당 부분 null로 수정하였습니다

yxxnhx committed:
RestaurantsInfoPage에서 관심사에 따른 컴포넌트 분리

페이지에서는 params로 라우팅만 할 수 있도록 관심사를 분리하였습니다

yxxnhx committed:
RestaurantsInfoContainer 컴포넌트 관심사에 따른 분리

컨테이너에서는 리덕스에서 값을 받아와서 로딩만 해줄 수 있도록 관심사를 분리하여 정리하였습니다

yxxnhx committed:
RestaurantsInfoContainer 테스트 관심사 분리에 따른 테스트 수정

페이지에서 id를 받아와 그에 해당하는 데이터를 리덕스에서 빼오는 것으로 관심사를 분리하여 id에 맞는 데이터를 받을 수 있도록 테스트를 수정하였습니다

yxxnhx committed:
RestaurantDetail 컴포넌트 생성 및 관심사 분리

디테일에서는 레스토랑에 대한 기본 정보만 출력할 수 있도록 관심사를 분리하여 생성하였습니다

yxxnhx committed:
RestaurantDetail 테스트 파일 생성 및 구현

컨테이너에서 받아온 restaurantInfo에 맞게 레스토랑 정보가 랜더링되는지 여부를 테스트하였습니다

yxxnhx committed:
RestaurantMenuItem 컴포넌트 생성 및 구현

관심사 분리로 메뉴 아이템 컴포넌트에서는 메뉴 아이템만 디테일 컴포넌트에서 받아 화면에 출력할 수 있도록 하였습니다

yxxnhx committed:
RestaurantMenuItem 테스트 파일 생성 및 구현

메뉴 아이템 정보의 여부에 따른 테스트 조건을 추가하여 생성하였습니다

yxxnhx commented:
# 진행사항
restaurantInfo 초기값 null로 모두 통일하여 수정하였습니다.
그리고 presentational component를 만들 생각을 하지 못하였었어요..🥹
일단 구현하는데 급급해서 컨테이너 안에 그냥 다 넣어서 진행했었습니다
그래서 RestaurantsInfoPage에서는 id만,  RestaurantsInfoContainer에서는 리덕스를 이용하여 액션을 꺼내오기만,  RestaurantDetail에서는 레스토랑 정보 출력만,  RestaurantMenuItem에서는 메뉴 아이템 출력만 할 수 있도록 관심사를 분리하여 컴포넌트 분리하는 것으로 수정하였습니다.

그리고 어제 질문하였던 부분은 구글링을 하여서 상대경로 때문에 `/`가 없을 경우에는 에러가 나서 일어났던 깨짐으로 이해하였는데 혹시 맞을까요? 사실 완벽하게 이해는 하지 못하였습니다 
제가 참고할 수 있는 사이트나 검색어를 알려주신다면 정말 감사하겠습니다!

daadaadaah commented on src/RestaurantsInfoPage.jsx:
```diff
@@ -0,0 +1,11 @@
+import RestaurantsInfoContainer from './RestaurantsInfoContainer';
+
+export default function RestaurantsInfoPage() {
+  const { id } = useParams();
```
`RestaurantsInfoPage`에서 id가 필요하지 않으므로, 필요한 곳에서만 useParams로 뽑아 사용할 수 있을거예요!

yxxnhx replied:
아하 넵! 수정하겠습니다~!

daadaadaah reviewed (changes requested):
## 질문에 대한 답변
- 네, `절대경로`와 `상대경로`의 차이에 대해 공부해보시면되요!
- 추가로, `main.js`에 대한 문제는 `index.html`의 `head`에
`<base href="/">`를 추가해서 해결할 수도 있습니다.
- 참고 : https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

yxxnhx committed:
절대 경로를 위해 index.html의 기본 경로 / 추가

절대 경로를 위하여 base href를 사용하여 기본 경로 /를 지정하여 경로를 찾지 못하는 에러 이슈를 해결할 수 있도록 수정하였습니다

yxxnhx committed:
RestaurantsInfoPage 컴포넌트에서 id 삭제

직접 사용하는 컨테이너에서 사용할 수 있도록 삭제하였습니다

yxxnhx committed:
RestaurantsInfoContainer 컴포넌트에 id 추가

직접 필요한 id를 꺼내와서 바로 사용할 수 있도록 수정하였습니다

yxxnhx commented:
# 진행사항
절대경로 상대 경로의 문제였다니! 감사합니다
덕분에 이해가 되었습니다. 처음에는 뭔가 설정이 잘못된 줄 알고 webpack 문서도 찾아보고 그랬었어요😅
그리고 `RestaurantsInfoContaienr에서 직접 사용할 수 있도록 `RestaurantsInfoPage`에서 빼왔습니다.

```jsx
        <Routes>
          <Route path="/" element={<HomePage />} />
          <Route path="/about" element={<AboutPage />} />
          <Route path="/restaurants" element={<RestaurantsPage />} />
          <Route path="/restaurants/:id" element={<RestaurantsInfoContaienr />} />
          <Route path="*" element={<NotFoundPage />} />
        </Routes>
```
그렇게 되면 페이지에서는 그저 컨테이너를 넘기는 일만 하게 되는데 페이지를 삭제하고 컨테이너에서 바로 보일 수 있도록 App의 라우트 설정을 변경해도 괜찮을까요?
그게 아니라면 페이지를 그대로 살려두는 것이 맞나요? 뭔가 라우트를 페이지로 하는 것이 맞는거 같은데 그렇다고 아무 일도 하지 않고 컨테이너로만 넘기는 일을 하는 컴포넌트를 살려두는 것이 맞는지 잘 모르겠습니다ㅠㅠ

daadaadaah commented:
> # 진행사항
> 절대경로 상대 경로의 문제였다니! 감사합니다 덕분에 이해가 되었습니다. 처음에는 뭔가 설정이 잘못된 줄 알고 webpack 문서도 찾아보고 그랬었어요😅 그리고 `RestaurantsInfoContaienr`에서 직접 사용할 수 있도록 `RestaurantsInfoPage`에서 빼왔습니다.
> 
> ```js
>   <Routes>
>     <Route path="/" element={<HomePage />} />
>     <Route path="/about" element={<AboutPage />} />
>     <Route path="/restaurants" element={<RestaurantsPage />} />
>     <Route path="/restaurants/:id" element={<RestaurantsInfoContaienr />} />
>     <Route path="*" element={<NotFoundPage />} />
>   </Routes>
> ```
> 
> 그렇게 되면 페이지에서는 그저 컨테이너를 넘기는 일만 하게 되는데 페이지를 삭제하고 컨테이너에서 바로 보일 수 있도록 위처럼 App의 라우트 설정을 변경해도 괜찮을까요? 그게 아니라면 페이지를 그대로 살려두는 것이 맞나요? 뭔가 라우트를 페이지로 하는 것이 맞는거 같은데 그렇다고 아무 일도 하지 않고 컨테이너로만 넘기는 일을 하는 컴포넌트를 살려두는 것이 맞는지 잘 모르겠습니다ㅠㅠ

- 추상화 수준을 일치시키기 위해 Page 컴포넌트로 라우트를 해주는 게 더 좋을 것 같아요!
- 저라면, page 컴포넌트에서 load 해옴으로써 Page 컴포넌트의 역할을 부여해줄 것 같아요! 다른 page 컴포넌트에서도 초기 데이터 불러와주니까!
